src/iterators/windows.rs

```diff
@@ -147,3 +147,119 @@ impl_iterator! {
         }
     }
 }
+
+/// Window producer and iterable
+///
+/// See [`.axis_windows()`](ArrayBase::axis_windows) for more
+/// information.
+pub struct AxisWindows<'a, A, D>{
+    base: ArrayView<'a, A, D>,
+    axis_idx: usize,
+    window: D,
+    strides: D,
+}
+
+impl<'a, A, D: Dimension> AxisWindows<'a, A, D> {
+    pub(crate) fn new(a: ArrayView<'a, A, D>, axis: Axis, window_size: usize) -> Self
+    {   
+        let strides = a.strides.clone();
+        let mut base = a;
+        let axis_idx = axis.index();
+        let mut window = base.raw_dim();
+        window[axis_idx] = window_size;
+
+        base.slice_each_axis_inplace(|ax_desc| {
```

Comment:
I am not really happy about the code duplication between this method and `Windows::new_with_stride`. Is it reasonably possible to have re-use without making this too hard to follow? (`Windows::new_with_stride` also appears to have additional assertions/consistency checks?)

Reply:
> The struct GeneralWindow and AxisWindow need to be public, but they can be inaccessible from outside of the crate.

This does sound reasonable to me

I would suggest making `Windows` and `AxisWindows` public and have both wrap an `WindowsImpl` or something like that. (Meaning `AxisWindows` contains two fields, `index: usize` and `impl: WindowsImpl`.)

I would also suggest replacing the term "version" by "variant" to avoid any confusion as "version" often has a temporal connotation.

> I don't know if it is more or less clean.

Of course, if this just increases complexity for very little gain in the end, it might not be worth it, you just have to give it a try.

As an alternative, a simple helper function encapsulating the slice-in-place approach using references to the constituent fields of `Windows` and `AxisWindows` might work as well.

Reply:
I just updated to my suggested structure, It actually does reduce code duplication a lot.

Reply:
I am not a fan of exposing these implementation detail generics in the API, hence the suggestion to expose `Windows` and `AxisWindows` with the same generic parameters as before which contain a `WindowsImpl/Inner/Whatever` as a private field. We should not leak this sharing of implementation in the public API.

Reply:
I thought that _This does sound reasonable to [you]_ 🙁

We need to expose that there is a difference between the return type of `windows()` and `axis_windows()` there is no way around that.

This generic is not an implementation detail. It makes a real difference to the public API. Specifically it changes the behavior of `<Windows as NdProducer>` and the Type of `<Windows as NdProducer>::Dim`. Otherwise this PR would be pointless.

Now we have this solution which is as minimal as it gets. 
Or the solution I had before. Sure I could try and faff around with some helper methods and structs to reduce code duplication for the cost of clarity and more boilerplate.
Having a `WindowsInner/Impl/...` will break the `impl_ndproducer!` macro, because `base` needs to be a field of `Windows` not of some wrapped type.

Reply:
> We need to expose that there is a difference between the return type of windows() and axis_windows() there is no way around that.

Distinct return types is not the issue. It is return types which only differ in a type parameter which I have a problem with. I want the relation between these two return types to be fully opaque to the caller, so that e.g. we could also not have code reuse in their implementation without that affecting our API.

> This generic is not an implementation detail. It makes a real difference to the public API. Specifically it changes the behavior of <Windows as NdProducer> and the Type of <Windows as NdProducer>::Dim. Otherwise this PR would be pointless.

`<Windows as NdProducer>::Dim` and `<AxisWindows as NdProducer>::Dim` without the outside world seeing any relation between those two types.

> Having a WindowsInner/Impl/... will break the impl_ndproducer! macro, because base needs to be a field of Windows not of some wrapped type.

This is a real issue and maybe manually implementing `NdProducer` as you did for `AxisWindows` for both case is simple enough, maybe this actually reduces reuse. This is why I also suggest just using an external helper method with completely disjoint struct types.

Reply:
Thank you for clearing that up. This makes much more sense.

I just extracted the main logic of creating the base into a separate function, and call that from both constructors `Windwos::new_with_stride()` and `AxisWindows::new()`